libtransact/src/workload/runner.rs

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright 2021 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+//! Provides a runner to submit `BatchWorkload`s
+
+// TODO add builder for Worker
+
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::mpsc::{channel, Sender, TryRecvError};
+use std::{thread, time};
+
+use reqwest::{blocking::Client, header, StatusCode};
+
+use crate::protos::IntoBytes;
+
+use super::error::WorkloadRunnerError;
+use super::BatchWorkload;
+
+const DEFAULT_LOG_TIME: u32 = 5; // time in seconds
+
+/// Keeps track of the currenlty running workloads.
+///
+/// The `WorkloadRunner` enables running different workloads, against different targets at
+/// different rates without needing to set up multiple runners.
+#[derive(Default)]
+pub struct WorkloadRunner {
+    workloads: HashMap<String, Worker>,
+}
+
+impl WorkloadRunner {
+    /// Starts running a new workload
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - A unique ID for the workload
+    /// * `workload` - The `BatchWorkload` used to generate the batches that will be submitted
+    /// * `targets` - A list of URL for submitting the batches. The URL provided must be the full
+    ///              URL before adding `/batches` for submission
+    /// * `rate`- How many tranactions per second to submit
+    /// * `auth` - The string to be set in the Authorization header for the request
+    ///
+    /// Returns an error if a workload with that ID is already running or if the workload thread
+    /// could not be started
+    pub fn add_workload(
+        &mut self,
+        id: &str,
+        workload: Box<dyn BatchWorkload>,
+        targets: &[String],
```

Comment:
Does a workload generate transactions that are dependent on previous transactions? Would that be a problem if a single workload sends transactions to different targets that may not have the dependencies?

Reply:
I'm not sure that it's the workload generators concern.  It probably should be up to the targets to ensure that they are exchanging batches in a way such that the network as a whole has all of the dependencies (e.g. like how Sawtooth does it).